Export / import RSA to / from provider

levitte opened the pull request:
This lays the foundation for transferring RSA keys to / from providers.

It goes with the idea of using multiple OSSL_PARAM items with the same key to pass multiple primes, CRT exponents and CRT coefficients, and in doing so, simplifies it by making *p* and *q* part of the primes (they are the first two), *dP* and *dQ* part of the CRT exponents, and *qInv* part of the CRT coefficients.  Extra RSA functionality is added to support this.

- [x] documentation is added or updated (for the extra RSA functionality)
- [x] tests are added or updated (for the extra RSA functionality)

levitte committed:
Added functions for easy getting and setting all RSA parameters.

RSA_set0_all_params() is used to set all the primes, exponents and
coefficients.  RSA_get0_all_params() is used to get all the primes,
exponents and coefficients.

"All" includes p, q, dP, dQ and qInv without making them separate.

All arrays of numbers are implemented as stacks to make dynamic use
easier.

levitte committed:
New RSA keymgmt implementation to handle import / export of RSA keys

levitte commented:
The reason I'm making this a separate PR is because the same kind of work exists both in an upcoming key generation branch I'm working on and in @mattcaswell's #10152, and that part of the work should be synchronised, at least with regards to export/import.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

slontis commented on crypto/rsa/rsa_ameth.c:
```diff
@@ -1045,6 +1047,108 @@ static int rsa_pkey_check(const EVP_PKEY *pkey)
+
+    /* Public parameters must always be present */
+    if (n == NULL || e == NULL)
+        return NULL;
```
memory leak?

levitte replied:
Ah, thanks.  That should goto err.

slontis commented on crypto/rsa/rsa_ameth.c:
```diff
@@ -1045,6 +1047,108 @@ static int rsa_pkey_check(const EVP_PKEY *pkey)
+    params = ossl_param_bld_to_param(&tmpl);
+
+    /* We export, the provider imports */
+    provkey = evp_keymgmt_importkey(keymgmt, params);
```
It doesnt look like it will crash if params returns NULL - not sure it is good form passing it to importkey as NULL though

levitte replied:
Check added

slontis commented:
Keep in mind that in fips mode multiprimes are not supported.
see the #ifdef FIPS_MODE in rsa files.
e.g: rsa_builtin_keygen

levitte commented:
Sure, but that will be for key generation in the provider, no?  Or should the import functions do some extra checks?  But in that case, shouldn't the RSA functionality simply refuse working with multi-primes when FIPS_MODE is defined, going so far as disabling RSA_set0_multi_prime_params() completely?  That's not what happens now.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte commented:
Fixing the multiprime issue with FIPS is for another PR, methinks.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

mattcaswell commented on crypto/rsa/rsa_lib.c:
```diff
@@ -613,3 +617,116 @@ int RSA_pkey_ctx_ctrl(EVP_PKEY_CTX *ctx, int optype, int cmd, int p1, void *p2)
+        sk_RSA_PRIME_INFO_pop_free(old, rsa_multip_info_free);
+    }
+
+    r->version = RSA_ASN1_VERSION_MULTI;
```
If `pnum == 2` then this is wrong

mattcaswell commented on providers/implementations/keymgmt/rsa_kmgmt.c:
```diff
@@ -0,0 +1,177 @@
+        && !OSSL_PARAM_set_BN(p, rsa_d))
+        goto err;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_RSA_PARAMS_NUM)) != NULL
```
Is this actually necessary? We don't seem to use it in the equivalent `params_to_key` function

levitte replied:
It's gone now

levitte commented on doc/man3/RSA_get0_key.pod:
```diff
@@ -110,6 +117,14 @@ RSA object.
 
+RSA_set0_all_params() sets all primes, CRT exponents and CRT coefficients
+in the B<RSA> object I<r> to the contents of the stacks of BIGNUMs I<primes>,
+I<exps> and I<coeffs>.
```
Done

levitte committed:
test/keymgmt_internal_test.c: New test of keymgmt internals

This tests diverse internal KEYMGMT features.  The current existing
test checks that evp_keymgmt_export_to_provider() passes the key data
correctly through two instances of the default provider, and that the
resulting numbers at the end match the initial numbers.

levitte committed:
evp_keymgmt_export_to_provider(): adjust OSSL_PARAM array for transfer

It may be that the OSSL_PARAM array we used for getting parameter
values for a key had a few too many entries.  These are detected by
their return_size == 0.  Before making second export call, we prune
away these items so we only ask for parameters that exist.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

levitte commented:
I added an internal keymgmt tester, `test/keymgmt_internal_test.c`, which tests that pushing around an RSA key through two providers (thereby exercising both import and export of key data) works as intended (it retrieves the numbers from the last provider instance and compares them to the original numbers).

levitte commented on doc/man3/RSA_get0_key.pod:
```diff
@@ -43,6 +44,12 @@ RSA_set0_multi_prime_params, RSA_get_version
+                         STACK_OF(BIGNUM_const) *coeffs);
+ int RSA_set0_all_params(RSA *r, const STACK_OF(BIGNUM) *primes,
+                         const STACK_OF(BIGNUM) *exps,
+                         const STACK_OF(BIGNUM) *coeffs);
```
Do you insist on internalizing them?

mattcaswell replied:
Well, I may insist but I'm willing to hear arguments as to why we should make them public.

levitte replied:
Quite frankly, I find the other ones clumsy.  With those, you have to make two or three calls depending on the amount of primes:

``` C
RSA_set0_factors(rsa, p, q);
RSA_set0_crt_params(rsa, dP, dQ, qInv);
/* The following for multi-primes only */
RSA_set0_multi_prime_params(rsa, extra_primes, extra_exps, extra_coeffs);
```

when these new functions allow for just one call regardless of the number of primes et al:

``` C
RSA_set0_all_params(rsa, all_factors, all_exps, all_coeffs);
```

This is more like what an RSA key *is* outside of the ASN.1 constraints (the current calls are *very much* an effect of what needed to happen to the ASN.1 spec to make it backward compatible).

So most of all, my argument is convenience for our users.

I can see one thing that makes it less convenient, and it's the use of stacks.  I'm perfectly willing to make an array variant too (side note: it should be possible to extract an array from a stack, that would simplify so many things)

mattcaswell replied:
While I understand where you are coming from we should perhaps have considered doing it this way before we introduced the RSA_set0_multi_prime_params() functions et al. We should be seeking stability in our API as much as possible and also seeking to restrict its growth. Every time we add new functions that's another set of functions we have to support for a very long time. Sometimes we *have*  to introduce new functions because there is no other way. But that isn't the case with this. I think we should not add this to the public API at this stage. We can always reconsider later if there is sufficient demand and reason for it.

levitte commented:
Not WIP anymore

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte committed:
fixup! test/keymgmt_internal_test.c: New test of keymgmt internals

levitte commented:
> Let's see if a kick helps

Yup, it did

mattcaswell commented on test/keymgmt_internal_test.c:
```diff
@@ -0,0 +1,198 @@
+        || !TEST_ptr_ne(km1, km2))
+        goto err;
+
+    if (!TEST_ptr(evp_keymgmt_export_to_provider(pk, km1))
```
So we export this key, but don't do anything with the returned reference (not even free it!)?

levitte replied:
Nope.  It's cached in *pk*, so will be freed when *pk* is freed.

mattcaswell commented on test/keymgmt_internal_test.c:
```diff
@@ -0,0 +1,198 @@
+        0,                       /* C3 */
+        0                        /* Extra, should remain zero */
+    };
+    static unsigned long keydata[OSSL_NELEM(expected)] = { 0, };
```
Why is this static?

levitte replied:
To make it possible to construct *params* in compile time.

levitte commented on include/openssl/rsa.h:
```diff
@@ -201,6 +203,9 @@ extern "C" {
 # define RSA_get_app_data(s)             RSA_get_ex_data(s,0)
 
+/* Pre-declare for those who need it */
+STACK_OF(BIGNUM);
```
Ah, thanks.  I forgot to clean this away

levitte replied:
Done

levitte commented on doc/man3/RSA_get0_key.pod:
```diff
@@ -110,6 +117,16 @@ RSA object.
+RSA_get0_all_params() gets all primes, CRT exponents and CRT coefficients
+in the B<RSA> object I<r> and pushes them on the stacks of constant BIGNUMs
+I<primes>, I<exps> and I<coeffs>.  The B<RSA> object takes ownership of the
+BIGNUMs, but not of the stacks.
```
Yes, thanks

levitte replied:
Done

levitte commented on providers/implementations/keymgmt/rsa_kmgmt.c:
```diff
@@ -0,0 +1,248 @@
+                                OSSL_PKEY_PARAM_RSA_COEFFICIENT))
+            goto err;
+
+        /* It's ok if this private key just has n, e and d */
```
To do what, exactly?  The [Cryptographic Primitives in RFC 8017](https://tools.ietf.org/html/rfc8017#section-5) require n, e and d at a minimum.  "*c* = *m*<sup>*e*</sup> mod *n*" and "*m* = *c*<sup>*d*</sup> mod *n*" should ring a bell.  So the question is, what should dictate what we accept for a minimal RSA key, the mathematical principles, or the possibility that we have an implementation that can't deal without the CRT params ('cause the prime factors are not enough, as described in the [second form of RSADP](https://tools.ietf.org/html/rfc8017#section-5.1.2).  Does Postel's Law apply here?

mattcaswell replied:
You are right. It seems we are perfectly ok to operate without these values.

levitte commented on test/keymgmt_internal_test.c:
```diff
@@ -0,0 +1,198 @@
+    };
+    static unsigned long keydata[OSSL_NELEM(expected)] = { 0, };
+    OSSL_PARAM params[] = {
+        OSSL_PARAM_size_t(OSSL_PKEY_PARAM_RSA_N, &keydata[N]),
```
Ah, thanks

levitte replied:
Fixed

levitte commented on test/keymgmt_internal_test.c:
```diff
@@ -0,0 +1,198 @@
+        goto err;
+
+    if (!TEST_ptr(evp_keymgmt_export_to_provider(pk, km1))
+        || !TEST_ptr(provdata = evp_keymgmt_export_to_provider(pk, km2)))
```
Why?  After all, this is simply passing key payload between different providers.  Why does it matter that they are in different OpenSSL contexts?  *It's just numbers*

mattcaswell replied:
I mostly worry about applications doing this sort of thing. I envisage scenarios where there is confusion about what OPENSSL_CTX is to be used for any particular action, e.g. locks being acquired in one OPENSSL_CTX but then updating the global data held in another OPENSSL_CTX. The concept I had for OPENSSL_CTX was that they were mutually exclusive...and you couldn't pass objects created by one OPENSSL_CTX to be used by another. Maybe I'm worrying too much.

levitte replied:
But hold on, this isn't about passing around any possible object, this is about passing around *know*, *identifiable* and *shareable*. The whole point with this function was to be able to pass around an EVP_PKEY between implementations of the same algo. The library context is not related to all of this.

levitte replied:
That is to say, yes I think you worry too much 😉

richsalz replied:
FWIW, I think Matt's right and Richard's wrong.

levitte commented on crypto/rsa/rsa_lib.c:
```diff
@@ -613,3 +617,118 @@ int RSA_pkey_ctx_ctrl(EVP_PKEY_CTX *ctx, int optype, int cmd, int p1, void *p2)
+            BIGNUM *prime = sk_BIGNUM_value(primes, i);
+            BIGNUM *exp = sk_BIGNUM_value(exps, i);
+            BIGNUM *coeff = sk_BIGNUM_value(coeffs, i - 1);
+            /* Using rsa_multip_info_new() is wasteful, so allocate directly */
```
Done

levitte commented on providers/implementations/keymgmt/build.info:
```diff
@@ -3,10 +3,14 @@
 IF[{- !$disabled{dsa} -}]
   SOURCE[$DSA_GOAL]=dsa_kmgmt.c
 ENDIF
+IF[{- !$disabled{rsa} -}]
```
Also, `$DSA_GOAL` wasn't quite right, even though harmless *now* :wink: 

Fixed

levitte committed:
fixup! test/keymgmt_internal_test.c: New test of keymgmt internals

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte commented:
Most concerns answered, https://github.com/openssl/openssl/pull/10190#discussion_r335904628 unresolved, and rebased.

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

mattcaswell commented on crypto/rsa/rsa_lib.c:
```diff
@@ -613,3 +617,118 @@ int RSA_pkey_ctx_ctrl(EVP_PKEY_CTX *ctx, int optype, int cmd, int p1, void *p2)
+    return 0;
+}
+
+DEFINE_SPECIAL_STACK_OF_CONST(BIGNUM_const, BIGNUM)
```
And here.

levitte replied:
Same reasoning

levitte commented on crypto/rsa/rsa_lib.c:
```diff
@@ -613,3 +617,118 @@ int RSA_pkey_ctx_ctrl(EVP_PKEY_CTX *ctx, int optype, int cmd, int p1, void *p2)
      return EVP_PKEY_CTX_ctrl(ctx, -1, optype, cmd, p1, p2);
 }
+
+DEFINE_STACK_OF(BIGNUM)
```
That macro defines static inline functions, so it has to be used where the stack functionality is used.  Using this line in bn_lib.c will benefit bn_lib.c and bn_lib.c only.

mattcaswell replied:
ok

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

levitte committed:
fixup! test/keymgmt_internal_test.c: New test of keymgmt internals

levitte committed:
fixup! New RSA keymgmt implementation to handle import / export of RSA keys

levitte commented:
Rebased on a fresher master.  That's a moving target!

levitte committed:
fixup! Added functions for easy getting and setting all RSA parameters.

mattcaswell reviewed (approved):
Approved assuming the CHANGES entry is removed as suggested below.

levitte commented:
Merged.  Pfew!  Thanks for the thorough review!

c3a4fa4cb849ba300d4691154c0c12ac19d24710 Added internal functions for easy getting and setting all RSA parameters.
29be60235b9bf86bb38f28349b405bb112250d4b New RSA keymgmt implementation to handle import / export of RSA keys
5a02d13d3e274748c695bfe19238f885489e021e test/keymgmt_internal_test.c: New test of keymgmt internals
651101e18d66b2ae89851ce8906299e9d2a871e0 evp_keymgmt_export_to_provider(): adjust OSSL_PARAM array for transfer

richsalz commented:
To make sure I understand: in the non-multiprime case you have several stacks of one entry.
In the multi-prime case you have to iterate through all the stacks in parallel?

levitte commented:
In the 2-prime case, the factors or primes stack has two numbers (commonly known as *p* and *q*), the exps stack has two numbers has two numbers (commonly known as *dP* or *dmp1*, and *dQ* or *dmq1*), and the coeffs stack has one number (commonly known as *qInv* or *iqmp*). Whether you are building a 2-prime or a n-prime key, processing them in parallel is more practical because of the way the RSA structure is constructed (which, BTW, is directly related to the ASN.1 spec)

richsalz commented:
thanks.  dees coeffs have to be a stack?

levitte commented:
It's more practical that way, for where those functions are used.  For n-primes, there's more than one, and I see very little benefit in special-casing 2-prime keys.

richsalz commented:
I missed the "for n-primes" part.  Thanks again.
